libevm/lib/service_state.go

```diff
@@ -229,6 +229,14 @@ func (s *Service) StateSetCode(params CodeParams) error {
 	return nil
 }
 
+func (s *Service) StateGetRefund(params HandleParams) (error, hexutil.Uint64) {
+	err, statedb := s.statedbs.Get(params.Handle)
+	if err != nil {
+		return err, 0
+	}
+	return nil, (hexutil.Uint64)(statedb.GetRefund())
```

Comment:
Good question, it should be applicable to all message processors, but turns out currently it is not.

If operations on the state free up space some gas is returned as a kind of "refund". I thought this is tracked automatically by StateDB, but turns out that part is in the implementation of the EVM op-code for `SSTORE`. Our custom message processors use libevm and therefore directly access StateDB, circumventing that refund logic.

I think we should leave it for now and put a TODO to update the libevm methods accessing the storage to track the refund identically to SSTORE.

See SSTORE implementation:
https://github.com/ethereum/go-ethereum/blob/0ce494b60cd00d70f1f9f2dd0b9bfbd76204168a/core/vm/gas_table.go#L96
(of course there are two version, because there is another EIP that changed this behavior)

Reply:
oof, the gas consumption of `SSTORE` has also changed multiple times. Involved EIPS [2200](https://eips.ethereum.org/EIPS/eip-2200), [2929](https://eips.ethereum.org/EIPS/eip-2929), [3529](https://eips.ethereum.org/EIPS/eip-3529).

Reply:
ok, we should use the most actual EIP

Reply:
created a ticket for this EVM-275